subcommand/acl-controller/command.go

```diff
@@ -147,6 +162,101 @@ func (c *Command) Help() string {
 	return ""
 }
 
+// upsertConsulResources creates the necessary resources in Consul if they do not exist.
+// This includes the partition, client role and policy, client and service token auth methods,
+// and the necessary binding rules.
+func (c *Command) upsertConsulResources(consulClient *api.Client, ecsMeta awsutil.ECSTaskMeta) error {
+	account, err := ecsMeta.AccountID()
+	if err != nil {
+		return err
+	}
+	path := strings.Trim(c.flagIAMRolePath, "/")
+	if path != "" {
+		path += "/"
+	}
+	boundArnPattern := fmt.Sprintf("arn:aws:iam::%s:role/%s*", account, path)
+
+	clientAuthMethod := &api.ACLAuthMethod{
+		Name:        "iam-ecs-client-token",
+		Type:        "aws-iam",
+		Description: "AWS IAM auth method for ECS client tokens",
+		Config: map[string]interface{}{
+			// Trust a wildcard - any roles at a path.
+			"BoundIAMPrincipalARNs": []string{boundArnPattern},
+			// Must be true to use a wildcard
+			"EnableIAMEntityDetails": true,
+		},
+	}
+	serviceAuthMethod := &api.ACLAuthMethod{
+		Name:        "iam-ecs-service-token",
+		Type:        "aws-iam",
+		Description: "AWS IAM auth method for ECS service tokens",
+		Config: map[string]interface{}{
+			// Trust a wildcard - any roles at a path.
+			"BoundIAMPrincipalARNs": []string{boundArnPattern},
+			// Must be true to use wildcard and tags
+			"EnableIAMEntityDetails": true,
+			"IAMEntityTags": []string{
+				authMethodServiceNameTag,
+				authMethodNamespaceTag,
+			},
+		},
+	}
+	if c.flagPartitionsEnabled {
+		serviceAuthMethod.NamespaceRules = append(serviceAuthMethod.NamespaceRules,
+			&api.ACLAuthMethodNamespaceRule{
+				Selector:      fmt.Sprintf(`entity_tags["%s"] != ""`, authMethodNamespaceTag),
+				BindNamespace: fmt.Sprintf(`${entity_tags.%s}`, authMethodNamespaceTag),
+			},
+		)
+	}
+
+	roleName := "consul-ecs-client-role"
+	policyName := "consul-ecs-client-policy"
```

Comment:
The static role and policy names are okay. If there are multiple controllers in the same datacenter, they can share the same policy. If there are partitions, we'd end up with one role+policy per partition (one per controller), which is fine and necessary.